### PR TITLE
added implementation for fleet workflow

### PR DESCRIPTION
**Pull request by audrastump**

### CODE REVIEW

File: src/run.ts [hunk 1]
```diff
@@ -50,20 +51,37 @@ export async function run() {
...
-         clusterName,
-         '-f',
-         kubeconfigPath
-      ]
-      if (subscription) cmd.push('--subscription', subscription)
-      if (admin) cmd.push('--admin')
-      if (publicFqdn) cmd.push('--public-fqdn')
+      if (fleetName){
+         cmd = [
+            'fleet',
```
Comment (nwnt) [inline src/run.ts]: Can we do something like this
```typescript
cmd = [
  (fleetName)? 'fleet' : 'aks',
  ...
```

And then lines 81 and 82 can be:
```typescript
if (!fleetName) {
  if (admin) cmd.push ('--admin')
  if (publicFqdn) cmd.push ('--public-fqdn')
}
```

I'm not an expert in TS but might be worth a try.

Comment (audrastump) [reply src/run.ts]: good catch, will do that strategy with fleet and cluster name as well!

File: src/run.ts [hunk 2]
```diff
@@ -28,7 +28,8 @@ export async function run() {
       const resourceGroupName = core.getInput('resource-group', {
          required: true
       })
-      const clusterName = core.getInput('cluster-name', {required: true})
```
Comment (jim-minter) [inline src/run.ts:31]: recommend not removing `cluster-name` as that will impose burden on users when they upgrade the github action version.  If we really want to do this, we can do this in a separate PR.

Comment (jim-minter) [inline src/run.ts:31]: File mode for both these files has changed from 0644 to 0755 - please revert that change.

File: src/run.ts [hunk 3]
```diff
@@ -28,7 +28,8 @@ export async function run() {
...
+      const resourceName = core.getInput('resource-name', {required: true})
+      const resourceType = core.getInput('resource-type') || ''
```
Comment (jim-minter) [inline src/run.ts]:
```suggestion
      const resourceType = core.getInput('resource-type') || 'Microsoft.ContainerService/managedClusters'
```

File: src/run.ts [hunk 4]
```diff
@@ -37,6 +38,8 @@ export async function run() {
       const publicFqdnInput = core.getInput('public-fqdn') || ''
       const publicFqdn = publicFqdnInput.toLowerCase() === 'true'
 
+
+
```
Comment (jim-minter) [inline src/run.ts:58]: please check your PRs for unnecessary changes and remove them :-)

File: src/run.ts [hunk 5]
```diff
@@ -50,20 +53,25 @@ export async function run() {
...
-         resourceGroupName,
-         '--name',
-         clusterName,
-         '-f',
-         kubeconfigPath
-      ]
-      if (subscription) cmd.push('--subscription', subscription)
-      if (admin) cmd.push('--admin')
-      if (publicFqdn) cmd.push('--public-fqdn')
+            resourceType=='fleet' ? 'fleet':'aks',
```
Comment (jim-minter) [inline src/run.ts]:
```suggestion
            resourceType.tolower() == 'microsoft.containerservice/fleets' ? 'fleet' : 'aks',
```

^ something like this?

File: src/run.ts (hunk 3 above)
Comment (jim-minter) [reply src/run.ts]: also add some validation that will fail if resourceType isn't recognised - i.e. is not a lower case match against microsoft.containerservice/{fleets,managedclusters} for now

File: src/run.ts (hunk 5 above)
Comment (jim-minter) [reply src/run.ts]: nit: please leave the indentation as it was

**Review by jim-minter — CHANGES_REQUESTED** [review]
This is a great start!  Please take a look at the proposed changes and also revise run.test.ts in their light?

File: src/run.ts (hunk 2 above)
Comment (nwnt) [reply src/run.ts:31]: how about we keep `cluster-name` while also adding `resource-name`?  If `cluster-name` is supplied, we fall back to AKS behavior like before.  If it's not there, we use the new way with `resource-name` and `resource-type`.

File: src/run.ts [hunk 6]
```diff
@@ -52,18 +63,20 @@ export async function run() {
...
-         clusterName,
+         resourceName,
          '-f',
          kubeconfigPath
       ]
       if (subscription) cmd.push('--subscription', subscription)
-      if (admin) cmd.push('--admin')
-      if (publicFqdn) cmd.push('--public-fqdn')
+      if (resourceType!='fleet') {
+         if (admin) cmd.push('--admin')
```
Comment (jim-minter) [inline src/run.ts]: ideally should probably validate above against user setting (admin || publicFqdn) && fleet

File: src/run.ts [hunk 7]
```diff
@@ -28,7 +28,8 @@ export async function run() {
       const resourceGroupName = core.getInput('resource-group', {
          required: true
       })
-      const clusterName = core.getInput('cluster-name', {required: true})
+      const resourceName = core.getInput('cluster-name', {required: true})
```
Comment (jim-minter) [inline src/run.ts]: suggest avoiding the variable name change both here and in the unit tests

File: src/run.ts (hunk 6 above)
Comment (jim-minter) [reply src/run.ts]: Reactivating.  What I mean is, up above by `// check resource type is recognized`, I think you should also validate:

```js
if (resourceType === 'microsoft.containerservice/fleets' && admin) {
    throw Error('admin must not be true when resource type is Microsoft.ContainerService/fleets')
}
if (resourceType === 'microsoft.containerservice/fleets' && publicFqdn) {
    throw Error('public-fqdn must not be true when resource type is Microsoft.ContainerService/fleets')
}
```

Then you don't need any changes in this part of the file.

This way, an invalid configuration is specifically rejected early, which helps the customer, rather than ignoring it, which could confuse them.

File: src/run.ts [hunk 8]
```diff
@@ -37,6 +38,24 @@ export async function run() {
...
+         resourceType.toLowerCase() !== 'microsoft.containerservice/fleets' &&
+         resourceType.toLowerCase() !== 'microsoft.containerservice/managedclusters'
+      ) {
+         throw Error(
+            'Resource type not recognized, either Microsoft.ContainerService/managedClusters or Microsoft.ContainerService/fleets is valid'
+         )
+      }
+
+      //validate user is not using admin or publicFqdn flags with fleet resource
+      if (resourceType === 'microsoft.containerservice/fleets' && admin) {
```
Comment (jim-minter) [inline src/run.ts:52]: use toLowerCase() here and below

Comment (jim-minter) [reply src/run.ts:52]: You'd probably be easiest writing something like `const resourceType = (core.getInput('resource-type') || 'Microsoft.ContainerService/managedClusters').toLowerCase()` and remove it from everywhere else?

Comment (jim-minter) [reply src/run.ts:52]: (and do the same in the k8s-deploy PR)
